### PR TITLE
Center banner with properties and combine text with image

### LINKED ISSUE #197: Resize hook affects all frames
When using dashboard and emacsclient, resizing an emacs frame triggers the resize hook and all frames, making some frames uncentered.

Before:

2019-11-11-131543_1920x1080_scrot

After:

2019-11-11-131552_1920x1080_scrot

### LINKED ISSUE #386: Center content before GUI loads
Currently, the dashboard is first shown left-aligned and then a short moment later the content snaps to the center, if configured. While not a functional problem, this is pretty jarring.

I was wondering whether this is expected behaviour in the first place (maybe there's something wrong with my config), and if so whether this is fixable.

I've tried moving different config settings to early-init.el, but I didn't manage to fix it.

### LINKED ISSUE #388: Better approach for emacsclient
The following information from the README is suboptimal:

> In addition to the above, configure `initial-buffer-choice` to show Dashboard in frames created with `emacsclient -c` as follows:
> 
> ```elisp
> (setq initial-buffer-choice (lambda () (get-buffer-create "*dashboard*")))
> ```

This is suboptimal because the first time after `emacs --bg-daemon` that a frame is created, it will not properly load the dashboard with all the graphics:

image

and the user have to reload the config to get it fixed. This is most probably because the `initial-buffer-choice` is run before any frame exists, and therefore everything is in text-only mode.

A better way would be to have

```elisp
(add-hook 'server-after-make-frame-hook (lambda()
    (switch-to-buffer dashboard-buffer-name)
    (dashboard-mode)
    (dashboard-insert-startupify-lists)
    (dashboard-refresh-buffer)))
```

which uses [a proper hook](https://www.gnu.org/software/emacs/manual/html_node/elisp/Creating-Frames.html#index-server_002dafter_002dmake_002dframe_002dhook) and also does not disturb the `emacsclient --tty --create-frame`. The content of the lambda is inspired from https://github.com/emacs-dashboard/emacs-dashboard/issues/236#issuecomment-711360926 

I suggest removing this section from README (or move it to some other section) and instead extend the `dashboard-setup-startup-hook` function to contain the hook above as well.

If you agree with this suggestion, I can create a PR.

### PR DESCRIPTION
**Pull request by Stebalien**
This patch contains two (entangled) fixes:

1. Center the banner with text properties.
2. Display both the image and text banner at the same time, letting the frame pick the right one for the display type.

The first change, centering with text properties, ensures that the text is smoothly centered without flickering, even if the dashboard is displayed in multiple windows at the same time (each window will center independently).

I'm inserting both the text and image banner at the same time to ensure that we always get the correct banner. Before this change, creating the dashboard buffer before creating a GUI frame would display a text banner.

Known issue: This patch still has to pick whether to center by the image or by the text banner. It does this at redisplay time using a display conditional, but it gets a bit confused when switching back and forth between a terminal frame and a GUI frame.

fixes #386
fixes #197
related #388 (icons still aren't handled)

### CODE REVIEW

**Comment by jcs090218** [comment]
Cool, I think this patch would also resolve issue #197?

I am currently having a disability to work with my PC. Would someone like to test and review this patch? 

Just randomly pining some maintainers here, @mveytsman @ricardoricho

**Comment by Stebalien** [comment]
> Cool, I think this patch would also resolve issue https://github.com/emacs-dashboard/emacs-dashboard/issues/197?

Yes, it should fix that as well (added to the list).

**Comment by Stebalien** [comment]
It may also fix https://github.com/emacs-dashboard/emacs-dashboard/issues/292?

**Comment by JesusMtnez** [comment]
I have enabled the CI for the PR, but I can not tested locally. I have reviewed the code and looks good to me. If it is OK for you, @jcs090218 @ricardoricho , we could merge it. Feel free to approve if you agree with it.

File: dashboard.el
Comment (ricardoricho) [inline dashboard.el:408]: I think we don't need to restore the `recentf-list` to it's original value (also I think there is a typo to fix in `origial-recentf-list`) It's done at the end of the function.

File: dashboard.el
Comment (ricardoricho) [inline dashboard.el:428]: Here we are restoring the `recentf-list`

File: dashboard.el
Comment (ricardoricho) [inline dashboard.el:398]: Maybe we can named this `lambda` to `dashboard-items-list` or something more descriptive, I think is an important function and deserves a name 😄

File: dashboard-widgets.el
Comment (ricardoricho) [inline dashboard-widgets.el:216]: This is a new option, Should we add an example on the `README` file?

**Review by ricardoricho — COMMENTED** [review]
Great PR, I just have a few comments but everything looks good. The only doubt I have is about the `recentf-list`.

File: dashboard.el
Comment (Stebalien) [reply dashboard.el:408]: I'm preserving the current behavior here. I _believe_ this exists so that if some section (a) modifies the recentf list, some later section (b) won't be affected. E.g., if the recent files section comes after the org agenda section.

File: dashboard.el
Comment (Stebalien) [reply dashboard.el:398]: This lambda is currently a closure that reads/modifies the outer state.

File: dashboard-widgets.el
Comment (Stebalien) [reply dashboard-widgets.el:216]: done

File: dashboard.el
Comment (ricardoricho) [reply dashboard.el:408]: Ok, we can leave it that way and maybe later change the recent files behavior in a way that `dashboard-insert-startupify-lists` doesn't know anything about recent files.

I didn't think about each section not been affected, I just thought the "origial" `recentf-list` is saved before any dashboard section consume a file and polluted that list. After all the dashboard is done, we set back the `recentf-list` with  the "origial" list value like nothing happened.

File: dashboard.el
Comment (ricardoricho) [reply dashboard.el:398]: Ok, let's leave it as a "local" definition. But my point is that if a closure needs to read from the external state, it is better to have that value as a parameter and try to avoid "side effects" by not modifying the external scope.

Comment (Stebalien) [reply dashboard.el:398]: I generally agree (and didn't write this code, just moved it).

**Comment by JesusMtnez** [comment]
@ricardoricho @jcs090218 did you have any change to test it?

**Comment by jcs090218** [comment]
I would to like to test it before we merge this! I will probably do it next weekend (hopefully). If @ricardoricho tested and it works, feel free to merge it! ;)

**Review by jcs090218 — APPROVED** [review]

**Comment by seagle0128** [comment]
This PR introduced the error: `(void-function image-supported-file-p)`.
Emacs 28.2, on Windows and macOS.

**Comment by jcs090218** [comment]
Ah... okay. I think the function `image-supported-file-p` only exists after `29.1`. @Stebalien Can you make a patch to handle this?

Why are you using the function `dashboard-center-line`? I don't think that's intended to be used externally.

**Comment by seagle0128** [comment]
@jcs090218 I used `dashboard-center-line` to customize some functionalities. It is not a big problem. `dashboard-insert-center` is the alternative. Please fix the issue of `image-supported-file-p`.

**Pull request merged**
